oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationProvider.java

```diff
@@ -84,15 +98,20 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		try {
 			jwt = this.jwtDecoder.decode(bearer.getToken());
 		} catch (JwtException failed) {
-			OAuth2Error invalidToken;
-			try {
-				invalidToken = invalidToken(failed.getMessage());
-			} catch ( IllegalArgumentException malformed ) {
-				// some third-party library error messages are not suitable for RFC 6750's error message charset
-				invalidToken = invalidToken("An error occurred while attempting to decode the Jwt: Invalid token");
+			OAuth2Error invalidToken = invalidToken(failed.getMessage());
+			throw new OAuth2AuthenticationException(invalidToken);
+		}
+
+		OAuth2TokenValidatorResult<OAuth2Error> result = this.jwtValidator.validate(jwt);
+		if ( !result.isSuccess() ) {
+			Collection<OAuth2Error> details = result.getDetails();
+			if ( details.size() == 1 ) {
+				throw new OAuth2AuthenticationException(details.iterator().next());
+			} else {
+				throw new OAuth2DetailedAuthenticationException(details.iterator().next(), details);
```

Comment:
That is the only purpose.

Open to suggestions.

What I want to do is propagate the data to the "view" layer (e.g. AccessDeniedHandler), so that the code doesn't make a rendering decision at this point. For example, given n errors, what should the BearerTokenAccessDeniedHandler put in `error_description` in the `WWW-Authenticate` header? Insofar as there doesn't appear to be a consensus among other resource servers, I figured I would make the data available so that users could customize how it was rendered.

The other simple option is to simply return the first error only. This has the benefit of a lot less code, but the sizable drawback that the user only receives one error at a time.

I also explored subclassing OAuth2Error with some kind of CompositeOAuth2Error, but it honestly felt a little clumsy.

Again, open to feedback.